### ISSUE TITLE
symbol problems with the larch server

### ISSUE THREAD

**Issue by bruceravel** [xraypy/xraylarch#99]
After a fit, this is a symbol: `fit.datasets[0].epsilon_k`.

If I ask the larch server for that symbol via the `get_data` command, nothing is returned (the XMLRPC version of an empty response).

If, however, I do this right after the fit: `foo = fit.datasets[0]`,  then ask the server for `foo.epsilon_k`, I get the expected return value (a list of epsilon values for each k-weight).  

Thus,  the problem seems to be with the `[0]`.

This isn't an immediate problem for Demeter -- Demeter gives another name to each of the `fit.datasets[N]`.  However, it's either a bug or I am misunderstanding the syntax of using the server for that kind of symbol.

**Comment by newville** [comment]
Hmm, weird....  will investigate, but it may take a couple days.

**Comment by bruceravel** [comment]
No rush.  Demeter seems to be working ok.  Although #98 is a `show()` stopper, it's not a show stopper.   (I'm hilarious!)

**Comment by newville** [comment]
I believe this is fixed with b9ae0a02824e3378b09a1a4fe0e249c34dda3c69.   As with the `show x` vs `show(x)` Issue (#98), the basic problem was that `get_data()` was only looking for actual symbols in the symbol table, and taking an element or slice with `something[0]` is an expression that needs to be evaluated.   The change in b9ae0a02824e3378b09a1a4fe0e249c34dda3c69 makes `get_data()` return the value of any expression, which is only slightly more complicated, but way better behavior.

**Comment by bruceravel** [comment]
Was it your intention to so dramatically change the server API?  

Before https://github.com/xraypy/xraylarch/commit/b9ae0a02824e3378b09a1a4fe0e249c34dda3c69, A&A were above the 95% level with Larch.  Now they are close to the 0% level.

I am happy to change Demeter to work with the new API, but I would like to know if you are done changing things.

**Comment by newville** [comment]
Um, no it wasn't.   I thought the change was only to the behavior of get_data().   Did I break something else?

**Comment by bruceravel** [comment]
I will document the changes and get back to you.

**Comment by bruceravel** [comment]
I understand the issue now.  It used to be the case that the client returned a reference to a hash containing this dictionary:
```python
{"__shape__": [511], "__class__": "Array", "value": [39622.2, 39720.2, 39729.2, ..., 6716.2], "__dtype__": "float64"}
```
now it returns a stringification of that dictionary.  That's fine.  Slightly less convenient (because a stringified python dictionary cannot be evaled as a perl hash without some string processing), but easily something I can work with.

You can leave it as is, but this sort of change certainly causes tsuris for Demeter.

**Comment by newville** [comment]
Actually, this code needs to change some anyway for other reasons....
I'll try to write some tests for this.  I think I'd like to try to have the
commandline interface use client/server, but that's a bit more work.

**Comment by bruceravel** [comment]
A quick hack got things more-or-less working with the stringifications, but neither A or A will do very much right now.  That's mostly seems to be a matter of handling undefined values sensibly, which turns out to be a bit different for references and evaled strings.  Shouldn't be too much work.  I also clearly need to write some tests for the server-to-perl linkage.

**Comment by bruceravel** [comment]
BTW, a client/server CLI is a good idea.

**Comment by newville** [comment]
Just to be clear, I definitely broke the get_data() by json-encoding the data before sending out, where it gets json encoded.... duh!   So,  I'll attempt to return it to the previous behavior.

No excuse for not having this better (much better) tested.

**Comment by newville** [comment]
I now believe this is doing the right thing, and that you can ask `get_data()` for any expression.   Well, OK, that's not entirely true.  There are probably many data types that `get_data()` will barf on, but any *normal* data Larch type (float, int, string, list, tuple, complex number, dictionary, numpy array, Parameter, Group, or nested combinations thereof) can be returned.     

Since JSON itself doesn't support all these types in full (like distinction between tuple and list, or data type of shape for numpy arrays), what is *actually* returned by `get_data()` may need to be "re-decoded".  It was like this before for arrays, but now it works better, includes complex numbers and arrays as well as Parameters and Groups, and these can be nested. 

Basically, for any complicated data type, you'll get back a dictionary with a '__class__' element, which will be one of 'List', 'Tuple', 'Dict', 'Complex', 'Array', 'Parameter', or 'Group'.    For these, you'd need to recreate the object from the other dictionary elements.    The Python code for this step is at https://github.com/xraypy/xraylarch/blame/master/lib/utils/jsonutils.py#L58  I'd tried to keep this simple, but may have failed...

I thought it might make sense to ask for a Parameter, and that perhaps you have a  Parameter-like thing in Perl already.   Similarly, I thought it might make sense to ask for a Group at a time, and then decode the results in one call.

**Comment by bruceravel** [comment]
FWIW, I think all of Athena is again working with Larch as of https://github.com/bruceravel/demeter/commit/ff6d8d3967390ea227c69b7171608f44dc5f62f8

The parts of Hephaestus that need (ifeffit|larch) seem to work as well. Artemis is likley pretty spotty at the moment.

**Comment by newville** [comment]
Whew!   Sorry for breaking that so badly... and great to hear that Athena and Hephaestus both work!

**Comment by bruceravel** [comment]
I believe this issue is resolved.
